modules/javafx.graphics/src/main/java/com/sun/prism/mtl/MTLTexture.java

```diff
@@ -0,0 +1,236 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.prism.mtl;
+
+import com.sun.prism.MediaFrame;
+import com.sun.prism.PixelFormat;
+import com.sun.prism.Texture;
+import com.sun.prism.impl.BaseTexture;
+
+import java.nio.Buffer;
+import java.nio.ByteBuffer;
+import java.nio.FloatBuffer;
+import java.nio.IntBuffer;
+
+public class MTLTexture<T extends MTLTextureData> extends BaseTexture<MTLTextureResource<T>> {
+
+    private final MTLContext context;
+    private long texPtr;
+
+    MTLTexture(MTLContext context, MTLTextureResource<T> resource,
+               PixelFormat format, WrapMode wrapMode,
+               int physicalWidth, int physicalHeight,
+               int contentX, int contentY, int contentWidth, int contentHeight,
+               boolean useMipmap) {
+
+        super(resource, format, wrapMode,
+              physicalWidth, physicalHeight,
+              contentX, contentY, contentWidth, contentHeight, useMipmap);
+        this.context = context;
+        texPtr = resource.getResource().getResource();
+    }
+
+    MTLTexture(MTLContext context, MTLTextureResource<T> resource,
+               PixelFormat format, WrapMode wrapMode,
+               int physicalWidth, int physicalHeight,
+               int contentX, int contentY, int contentWidth, int contentHeight,
+               int maxContentWidth, int maxContentHeight, boolean useMipmap) {
+
+        super(resource, format, wrapMode,
+              physicalWidth, physicalHeight,
+              contentX, contentY, contentWidth, contentHeight,
+              maxContentWidth, maxContentHeight, useMipmap);
+        this.context = context;
+        texPtr = resource.getResource().getResource();
+    }
+
+    public long getNativeHandle() {
+        return texPtr;
+    }
+
+    public MTLContext getContext() {
+        return context;
+    }
+
+    // We don't handle mipmap in shared texture yet.
+    private MTLTexture(MTLTexture sharedTex, WrapMode newMode) {
+        super(sharedTex, newMode, false);
+        this.context = sharedTex.context;
+        this.texPtr = sharedTex.texPtr;
+    }
+
+    @Override
+    protected Texture createSharedTexture(WrapMode newMode) {
+        return new MTLTexture(this, newMode);
+    }
+
+    @Override
+    public void update(Buffer buffer, PixelFormat format,
+                        int dstx, int dsty,
+                        int srcx, int srcy,
+                        int srcw, int srch,
+                        int srcscan, boolean skipFlush) {
+
+        switch (format.getDataType()) {
```

Comment:
I would use an exhaustive switch and possibly extract each branch into a method. This is a big switch with sub-switches, it's easy to get lost.